halo2/src/prover.rs

```diff
@@ -17,29 +17,45 @@ use polyexen::plaf::PlafDisplayBaseTOML;
 use rand::{rngs::StdRng, SeedableRng};
 
 use crate::circuit_builder::analyzed_to_circuit;
+use std::io;
 
 /// Create a halo2 proof for a given PIL, fixed column values and witness column values
 /// We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and Keccak256
-pub fn prove_ast<T: FieldElement>(
+
+pub fn prove_ast_read_params<T: FieldElement, R: io::Read>(
     pil: &Analyzed<T>,
     fixed: Vec<(&str, Vec<T>)>,
     witness: Vec<(&str, Vec<T>)>,
+    mut params: R,
 ) -> Vec<u8> {
     if polyexen::expr::get_field_p::<Fr>() != T::modulus().to_arbitrary_integer() {
         panic!("powdr modulus doesn't match halo2 modulus. Make sure you are using Bn254");
     }
 
-    let circuit = analyzed_to_circuit(pil, fixed, witness);
+    prove_ast(
+        pil,
+        fixed,
+        witness,
+        ParamsKZG::<Bn256>::read(&mut params).unwrap(),
+    )
+}
 
-    let circuit_row_count_log = usize::BITS - circuit.plaf.info.num_rows.leading_zeros();
+pub fn prove_ast<T: FieldElement>(
+    pil: &Analyzed<T>,
+    fixed: Vec<(&str, Vec<T>)>,
```

Comment:
use alias, eventually a struct would be better

Reply:
removed alias